sale_order_product_recommendation/wizards/sale_order_recommendation.py

```diff
@@ -0,0 +1,198 @@
+            ["product_id"],
+        )
+        # Manual ordering that circumvents ORM limitations
+        found_lines = sorted(
+            found_lines,
+            key=lambda res: (
+                res["product_id_count"],
+                res["qty_delivered"],
+            ),
+            reverse=True,
+        )
+        found_dict = {l["product_id"][0]: l for l in found_lines}
+        RecomendationLine = self.env["sale.order.recommendation.line"]
+        existing_product_ids = []
+        # Add products from sale order lines
+        for line in self.order_id.order_line:
+            found_line = found_dict[line.product_id.id]
+            new_line = RecomendationLine.new({
+                "product_id": line.product_id.id,
+                "times_delivered": found_line["product_id_count"],
+                "units_delivered": found_line["qty_delivered"],
+                "units_included": line.product_uom_qty,
+                "sale_line_id": line.id,
+            })
+            self.line_ids += new_line
+            existing_product_ids.append(line.product_id.id)
+        # Add those recommendations too
+        i = 0
+        for line in found_lines:
+            if line["product_id"][0] in existing_product_ids:
+                continue
+            new_line = RecomendationLine.new({
+                "product_id": line["product_id"][0],
+                "times_delivered": line["product_id_count"],
+                "units_delivered": line["qty_delivered"],
+            })
+            self.line_ids += new_line
+            # limit number of results. It has to be done here, as we need to
+            # populate all results first, for being able to select best matches
+            i += 1
+            if i == self.line_amount:
+                break
+
+    @api.multi
+    def action_accept(self):
+        """Propagate recommendations to sale order."""
+        so_lines = self.env["sale.order.line"]
+        sequence = max(self.order_id.mapped('order_line.sequence') or [0])
+        for wiz_line in self.line_ids.filtered('is_modified'):
+            # Use preexisting line if any
+            if wiz_line.sale_line_id:
+                if wiz_line.units_included:
+                    wiz_line.sale_line_id.update({
+                        "product_uom_qty": wiz_line.units_included,
+                    })
+                    wiz_line.sale_line_id.product_uom_change()
+                else:
+                    wiz_line.sale_line_id.unlink()
+                continue
+            sequence += 1
+            # Use a new in-memory line otherwise
+            so_line = so_lines.new({
+                "order_id": self.order_id.id,
+                "product_id": wiz_line.product_id.id,
+                "sequence": sequence,
+            })
+            so_line.product_id_change()
+            so_line.product_uom_qty = wiz_line.units_included
+            so_line.product_uom_change()
+            so_lines |= so_line
```

Comment:
For performance is better store createed line id and finally do a browse with all ids...

Reply:
Even with on memory records? Changed anyway